File: dlox/lib/src/scanner.dart
import 'package:charcode/ascii.dart';

import 'exception.dart';
import 'nil.dart';
import 'token.dart';
import 'util.dart';

const _keywords = <String, TokenType>{
  'and': TokenType.AND,
  'class': TokenType.CLASS,
  'else': TokenType.ELSE,
  'false': TokenType.FALSE,
  'for': TokenType.FOR,
  'fun': TokenType.FUN,
  'if': TokenType.IF,
  'nil': TokenType.NIL,
  'or': TokenType.OR,
  'print': TokenType.PRINT,
  'return': TokenType.RETURN,
  'super': TokenType.SUPER,
  'this': TokenType.THIS,
  'true': TokenType.TRUE,
  'var': TokenType.VAR,
  'while': TokenType.WHILE,
};

class Scanner {
  final String _source;

  final _tokens = <Token>[];

  int _idx = 0;
  int _start = 0;
  int _line = 1;

  Scanner(this._source);

  bool get _isAtEnd => _idx >= _source.length;

  List<Token> scan() {
    while (!_isAtEnd) {
      _start = _idx;
      _scanToken();
    }
    _tokens.add(Token(TokenType.EOF, '', Nil(), _line));
    return _tokens;
  }

  void _scanToken() {
    final c = _advance();

    switch (c) {
      case $lparen:
        _addToken(TokenType.LEFT_PAREN);
        break;
      case $rparen:
        _addToken(TokenType.RIGHT_PAREN);
        break;
      case $lbrace:
        _addToken(TokenType.LEFT_BRACE);
        break;
      case $rbrace:
        _addToken(TokenType.RIGHT_BRACE);
        break;
      case $comma:
        _addToken(TokenType.COMMA);
        break;
      case $dot:
        _addToken(TokenType.DOT);
        break;
      case $minus:
        _addToken(TokenType.MINUS);
        break;
      case $plus:
        _addToken(TokenType.PLUS);
        break;
      case $semicolon:
        _addToken(TokenType.SEMICOLON);
        break;
      case $asterisk:
        _addToken(TokenType.STAR);
        break;
      case $exclamation:
        _addToken(_match($equal) ? TokenType.BANG_EQUAL : TokenType.BANG);
        break;
      case $equal:
        _addToken(_match($equal) ? TokenType.EQUAL_EQUAL : TokenType.EQUAL);
        break;
      case $less_than:
        _addToken(_match($equal) ? TokenType.LESS_EQUAL : TokenType.LESS);
        break;
      case $greater_than:
        _addToken(_match($equal) ? TokenType.GREATER_EQUAL : TokenType.GREATER);
        break;
      case $slash:
        if (_match($slash)) {
          while (_peek() != $lf && !_isAtEnd) {
            _advance();
          }
        } else {
          _addToken(TokenType.SLASH);
        }
        break;
      case $space:
      case $cr:
      case $tab:
        break;
      case $lf:
        _line++;
        break;
      case $double_quote:
        _string();
        break;
      default:
        if (isDigit(c)) {
          _number();
        } else if (isAlpha(c)) {
          _identifier();
        } else {
          _error(_line, 'Unexpected character');
        }
        break;
    }
  }

  void _string() {
    while (_peek() != $double_quote && !_isAtEnd) {
      if (_peek() == $lf) {
        _line++;
      }
      _advance();
    }

    if (_isAtEnd) _error(_line, 'Unterminated string.');

    _advance();
    _addToken(TokenType.STRING, _source.substring(_start + 1, _idx - 1));
  }

  void _number() {
    while (isDigit(_peek())) {
      _advance();
    }

    if (_peek() == $dot && isDigit(_peekNext())) {
      _advance();
    }

    while (isDigit(_peek())) {
      _advance();
    }

    _addToken(TokenType.NUMBER, double.parse(_source.substring(_start, _idx)));
  }

  void _identifier() {
    while (isAlphaNumeric(_peek())) {
      _advance();
    }

    _addToken(_keywords[_source.substring(_start, _idx)] ?? TokenType.IDENT);
  }

  int _advance() => _source.codeUnitAt(_idx++);

  void _addToken(TokenType type, [Object? literal]) {
    final text = _source.substring(_start, _idx);
    _tokens.add(Token(type, text, literal, _line));
  }

  bool _match(int expected) {
    if (_isAtEnd || (_source.codeUnitAt(_idx) != expected)) {
      return false;
    }
    _idx++;
    return true;
  }

  int _peek() {
    if (_isAtEnd) {
      return $nul;
    }
    return _source.codeUnitAt(_idx);
  }

  int _peekNext() {
    if (_idx + 1 >= _source.length) {
      return $nul;
    }

    return _source.codeUnitAt(_idx + 1);
  }

  void _error(int line, String message) {
    throw SyntaxError(line, message);
  }
}

File: rlox/src/scanner.rs
use std::fmt;
use std::str::{Chars, FromStr};

use crate::token::{Token, TokenKind};
use TokenKind::*;

#[derive(Debug, PartialEq, Copy, Clone)]
pub enum ScanError {
    UnexpectedChar(char),
    UnterminatedString,
}

impl fmt::Display for ScanError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ScanError::UnexpectedChar(_) => write!(f, "Unexpected character."),
            ScanError::UnterminatedString => write!(f, "Unterminated string."),
        }
    }
}

pub fn tokenize(mut src: &str) -> impl Iterator<Item = Token<'_>> {
    let mut start_line = 1;
    let mut at_end = false;

    std::iter::from_fn(move || {
        if at_end {
            return None;
        }

        if src.is_empty() {
            at_end = true;
            return Some(Token::new(EOF, start_line));
        }

        let (token, consumed) = Scanner::new(src, start_line).scan();
        start_line = token.line;
        src = &src[consumed..];

        Some(token)
    })
    .filter(|t| !t.is_whitespace())
}

struct Scanner<'a> {
    src: &'a str,
    chars: Chars<'a>,
    line: u64,
}

impl<'a> Scanner<'a> {
    fn new(src: &'a str, line: u64) -> Scanner<'a> {
        Scanner {
            src,
            chars: src.chars(),
            line,
        }
    }

    fn scan(&mut self) -> (Token<'a>, usize) {
        let c = self.advance().unwrap();

        let kind = match c {
            c if c.is_ascii_whitespace() => {
                self.advance_while(|c| c.is_ascii_whitespace());
                WHITESPACE
            }
            '/' => match self.peek() {
                Some('/') => {
                    self.advance_while(|c| c != '\n');
                    COMMENT
                }
                _ => SLASH,
            },
            c if is_ident_start(c) => self.identifier(),
            '0'..='9' => self.number(),
            '(' => LEFT_PAREN,
            ')' => RIGHT_PAREN,
            '{' => LEFT_BRACE,
            '}' => RIGHT_BRACE,
            ',' => COMMA,
            ';' => SEMICOLON,
            '.' => DOT,
            '-' => MINUS,
            '+' => PLUS,
            '*' => STAR,
            '!' => self.maybe_double(BANG, BANG_EQUAL),
            '=' => self.maybe_double(EQUAL, EQUAL_EQUAL),
            '>' => self.maybe_double(GREATER, GREATER_EQUAL),
            '<' => self.maybe_double(LESS, LESS_EQUAL),
            '"' => self.string(),
            _ => ERROR(ScanError::UnexpectedChar(c)),
        };

        (Token::new(kind, self.line), self.consumed())
    }

    fn identifier(&mut self) -> TokenKind<'a> {
        self.advance_while(|c| c.is_alphanumeric() || c == '_');
        let src = &self.src[..self.consumed()];
        TokenKind::from_str(src).unwrap_or(IDENTIFIER(src))
    }

    fn number(&mut self) -> TokenKind<'a> {
        self.advance_while(is_number);

        if let (Some('.'), Some(c)) = (self.peek(), self.peek_next()) {
            if is_number(c) {
                self.advance();
                self.advance_while(is_number)
            }
        }

        let src = &self.src[..self.consumed()];
        NUMBER(f64::from_str(src).unwrap())
    }

    fn string(&mut self) -> TokenKind<'a> {
        while let Some(c) = self.advance() {
            if c == '"' {
                return STRING(&self.src[1..self.consumed() - 1]);
            }
        }

        ERROR(ScanError::UnterminatedString)
    }

    fn maybe_double(&mut self, single: TokenKind<'a>, double: TokenKind<'a>) -> TokenKind<'a> {
        if let Some('=') = self.peek() {
            self.advance();
            return double;
        }

        single
    }

    fn advance(&mut self) -> Option<char> {
        self.chars.next().map(|c| {
            if c == '\n' {
                self.line += 1;
            };
            c
        })
    }

    fn advance_while(&mut self, mut f: impl FnMut(char) -> bool) {
        while let Some(c) = self.peek() {
            if !f(c) {
                break;
            }
            self.advance();
        }
    }

    fn peek(&self) -> Option<char> {
        self.chars.clone().next()
    }

    fn peek_next(&self) -> Option<char> {
        self.chars.clone().nth(1)
    }

    fn consumed(&self) -> usize {
        self.src.len() - self.chars.as_str().len()
    }
}

fn is_ident_start(c: char) -> bool {
    matches!(c, 'a'..='z' | 'A'..='Z' | '_')
}

fn is_number(c: char) -> bool {
    matches!(c, '0'..='9')
}
